src/libraries/Native/Unix/System.Native/pal_io.c

```diff
@@ -1100,10 +1106,114 @@ int32_t SystemNative_ReadLink(const char* path, char* buffer, int32_t bufferSize
     return (int32_t)count;
 }
 
-int32_t SystemNative_Rename(const char* oldPath, const char* newPath)
+// Potential issue: renaming a file onto another link to itself can decrease the total link count of the file.
+// We could only fix this if there were a way to ask the filesystem if two dnodes were identical, but there isn't.
+static int32_t RenameOnEexist(int32_t oldResult, struct stat *targetBuf, const char *oldPath, const char *newPath)
+{
+    struct stat oldBuf;
+    struct stat newBuf;
+    bool haveTarget = targetBuf != NULL;
+    if (!targetBuf)
+    {
+        targetBuf = &newBuf;
+    }
+    int savedErrno = errno;
+    errno = 0;
+    if (!lstat(oldPath, &oldBuf) && (haveTarget || !lstat(newPath, targetBuf))
+            && oldBuf.st_dev == targetBuf->st_dev && oldBuf.st_ino == targetBuf->st_ino)
+    {
+        errno = 0;
+        return rename(oldPath, newPath);
+    }
+    if (errno == EINTR)
+    {
+        return -1;
+    }
+    errno = savedErrno;
+    return oldResult;
+}
+
+int32_t SystemNative_Rename(const char* oldPath, const char* newPath, int32_t flags)
 {
     int32_t result;
-    while ((result = rename(oldPath, newPath)) < 0 && errno == EINTR);
+    if (flags & 1)
+    {
+        while ((result = rename(oldPath, newPath)) < 0 && errno == EINTR);
+    }
+    else
+    {
+#if HAVE_RENAMEAT2 && defined(RENAME_NOREPLACE)
+        // Linux renameat2() provides the functionality we want inline.
+        // This is in a feature check block as this API may well be ported to other OSes later.
+        do {
+            result = renameat2(AT_FDCWD, oldPath, AT_FDCWD, newPath, RENAME_NOREPLACE);
+            if (result >= 0)
+            {
+                return result;
+            }
+            else if (errno == EEXIST)
+            {
+                // But renameat2 returns EEXIST on renaming a file onto itself. Check that.
+                result = RenameOnEexist(result, NULL, oldPath, newPath);
+                if (result != EINTR)
+                {
+                    return result;
+                }
+            }
+            // Documentation says EINVAL is returned if the filesystem doesn't support it. We also check the standard codes for this.
+            else if (errno != EINTR && errno != EINVAL && errno != ENOSYS && errno != ENOTSUP && errno != EOPNOTSUPP
+#ifdef ENOTSUPP
+                 && errno != ENOTSUPP
+#endif
+            )
+            {
+                return result;
+            }
+        } while (errno == EINTR);
+        // If the underlying FS doesn't support the operation we end up here just as if we're not Linux.
+#endif
+        do {
+            result = -1; // Keeps compiler happy.
+            if (!(flags & 2))
+            {
+                result = link(oldPath, newPath);
+                if (result >= 0)
+                {
+                    while ((result = unlink(oldPath)) < 0 && errno == EINTR);
```

Comment:
Side effect: File system watcher looking for rename() won't ever see it.

Incidentally I had to go through our own codebase ten years ago and systematically swap File.Move() for P/Invoke of CreateHardLink because somebody deployed our application on Windows pointing at a NAS box and hit the race condition where Windows resolving a UNC to a NAS box will inherently have this race. There is one other side effect, that is if the application crashes, you can see both names at once. But it's not really a side effect on Unix, because *that can happen anyway* on power loss.